plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/LoggingInvoker.java

```diff
@@ -0,0 +1,165 @@
+        if (isInternal(clazz)) {
+            record(message.get());
+        }
+    }
+
+    private static @NonNull Class<?> classOf(@CheckForNull Object receiver) {
+        if (receiver == null) {
+            return Void.class;
+        } else if (receiver instanceof Class) {
+            return (Class<?>) receiver;
+        } else {
+            return receiver.getClass();
+        }
+    }
+
+    @Override public Object methodCall(Object receiver, String method, Object[] args) throws Throwable {
+        Class<?> clazz = classOf(receiver);
+        maybeRecord(clazz, () -> clazz.getName() + "." + method);
+        return delegate.methodCall(receiver, method, args);
+    }
+
+    @Override public Object constructorCall(Class lhs, Object[] args) throws Throwable {
+        maybeRecord(lhs, () -> lhs.getName() + ".<init>");
+        return delegate.constructorCall(lhs, args);
+    }
+
+    @Override public Object superCall(Class senderType, Object receiver, String method, Object[] args) throws Throwable {
+        maybeRecord(senderType, () -> senderType.getName() + "." + method);
+        return delegate.superCall(senderType, receiver, method, args);
+    }
+
+    @Override public Object getProperty(Object lhs, String name) throws Throwable {
+        Class<?> clazz = classOf(lhs);
+        maybeRecord(clazz, () -> clazz.getName() + "." + name);
+        return delegate.getProperty(lhs, name);
+    }
+
+    @Override public void setProperty(Object lhs, String name, Object value) throws Throwable {
+        Class<?> clazz = classOf(lhs);
+        maybeRecord(clazz, () -> clazz.getName() + "." + name);
+        delegate.setProperty(lhs, name, value);
+    }
+
+    @Override public Object getAttribute(Object lhs, String name) throws Throwable {
+        Class<?> clazz = classOf(lhs);
+        maybeRecord(clazz, () -> clazz.getName() + "." + name);
+        return delegate.getAttribute(lhs, name);
+    }
+
+    @Override public void setAttribute(Object lhs, String name, Object value) throws Throwable {
+        Class<?> clazz = classOf(lhs);
+        maybeRecord(clazz, () -> clazz.getName() + "." + name);
+        delegate.setAttribute(lhs, name, value);
+    }
+
+    @Override public Object getArray(Object lhs, Object index) throws Throwable {
+        return delegate.getArray(lhs, index);
+    }
+
+    @Override public void setArray(Object lhs, Object index, Object value) throws Throwable {
+        delegate.setArray(lhs, index, value);
+    }
+
+    @Override public Object methodPointer(Object lhs, String name) {
+        Class<?> clazz = classOf(lhs);
+        maybeRecord(clazz, () -> clazz.getName() + "." + name);
+        return delegate.methodPointer(lhs, name);
+    }
+
+    @Override public Object cast(Object value, Class<?> type, boolean ignoreAutoboxing, boolean coerce, boolean strict) throws Throwable {
+        Class<?> clazz = classOf(value);
+        maybeRecord(clazz, () -> clazz.getName() + " as " + type.getClass().getName());
```

Comment:
I would probably check `type` instead of `clazz` here (or check both), or just drop this entirely. If `clazz` is a Jenkins type, then we should have already logged it earlier via some method call or property access, so I think this is redundant, but if you are casting some ordinary object to a Jenkins type then if the sandbox is inactive the construction of that object might not be intercepted otherwise. Of course just instantiating the object is probably not interesting, and whatever you do with it is likely to be intercepted separately, so maybe its fine to just not check anything here.

Reply:
OK I can just skip it, probably a corner case.